text/0000-custom-dst.md

```diff
@@ -0,0 +1,480 @@
+    }
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        &mut self.buff as *mut [T; 0] as *mut T
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        unsafe {
+            std::slice::from_raw_parts(self.as_ptr(), self.len())
+        }
+    }
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        unsafe {
+            std::slice::from_raw_parts(self.as_mut_ptr(), self.len())
+        }
+    }
+
+    // panics if it doesn't have remaining capacity
+    pub fn push(&mut self, el: T) {
+        assert!(self.size() < self.capacity());
+        let ptr = self.as_mut_ptr();
+        let index = self.len();
+        std::ptr::write(ptr.offset(index as isize), el);
+        self.len += 1;
+    }
+
+    // panics if it doesn't have any elements
+    pub fn pop(&mut self) -> T {
+        assert!(self.len() > 0);
+        self.len -= 1;
+        let ptr = self.as_mut_ptr();
+        let index = self.len();
+        std::ptr::read(ptr.offset(index as isize))
+    }
+}
+```
+
+### Flexible Array Members
+
+Communicating with C types that contain flexible array members
+is an important part of this RFC.
+
+```rust
+// note: a real example from winapi
+#[repr(C)]
+struct TOKEN_GROUPS {
+    GroupCount: DWORD,
+    Groups: [SID_AND_ATTRIBUTES; 0],
+}
+
+impl Pointee for TOKEN_GROUPS {
+    type Metadata = ();
+}
+
+unsafe impl Contiguous for TOKEN_GROUPS {
+    fn size_of_val(&self) -> usize {
+        std::mem::size_of_header::<Self>()
+        + self.GroupCount * std::mem::size_of::<SID_AND_ATTRIBUTES>()
+    }
+
+    fn align_of_val((): ()) -> usize {
+        std::mem::align_of_header::<Self>()
+    }
+}
+
+extern "system" {
+    pub fn AdjustTokenGroups(
+        TokenHandle: HANDLE,
+        ResetToDefault: BOOL,
+        NewState: &mut TOKEN_GROUPS,
+        BufferLength: DWORD,
+        PreviousState: Option<&mut TOKEN_GROUPS>,
+        ReturnLength: &mut DWORD,
+    ) -> BOOL;
+}
+```
+
+### 2D Views of Planes
+
+A reasonably tiny example of a 2D view of a plane.
+This is less important for common Rust,
+but should be helpful for graphics programming, for example.
+
+```rust
+// owned Plane<T>
+struct PlaneBuf<T> {
+    stride: usize,
+    buffer: Box<[T]>,
+}
+
+impl<T> Deref for PlaneBuf<T> {
+    type Target = Plane<T>;
+
+    fn deref(&self) -> &Plane<T> {
+        let ptr = &*self.buffer;
+        let meta = PlaneMetadata {
+            width: self.stride,
+            stride: self.stride,
+            height: buffer.len() / width,
+        };
+
+        unsafe {
+            &*std::raw::from_raw_parts::<Plane<T>>(ptr, meta)
+        }
+    }
+}
+
+// borrowed Plane<T>
+struct Plane<T> {
+    buffer: [T; 0],
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct PlaneMetadata {
+    width: usize,
+    stride: usize,
+    height: usize,
+}
+
+unsafe impl<T> DynamicallySized for Plane<T> {
+    type Metadata = PlaneMetadata;
+}
+
+// note that `Contiguous` is not implemented for `Plane<T>`,
```

Comment:
What prevents the automatic implementation of `Contiguous` given that `struct Plane<T> { buffer: [T; 0] }` is `Sized`? Does implementing `Pointee` make the implementation of `Sized` go away?